### PR TITLE
Replace *Animation traits that use PlayFetchIndex with *SpriteBodies

### PR DESCRIPTION
**Pull request by reaperrr**
1) Replaces the two special-case `With*Animation` traits that use `PlayFetchIndex` -  `WithChargeAnimation` and `WithSiloAnimation` - with `*SpriteBody` equivalents akin to similar cases like `WithGateSpriteBody` or `WithWallSpriteBody`.
`PlayFetchIndex` isn't compatible with other animation traits (unless as overlay), so these cases are better left as special-case sprite bodies (all of the affected actors in our shipping mods either don't need generic animation support, or already use overlays).

2) Renames `WithResources` to `WithResourceLevelOverlay` and fixes TS silo yaml setup.

### CODE REVIEW

**Review by abcdefg30 — DISMISSED** [review]
Looks good to me.

File: OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	{
+		readonly AttackCharges attackCharges;
+
+		public WithChargeSpriteBody(ActorInitializer init, WithChargeSpriteBodyInfo info)
+			: base(init, info, () => 0)
+		{
+			attackCharges = init.Self.Trait<AttackCharges>();
+		}
+
+		protected override void TraitResumed(Actor self)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs]: Why is this on resumed rather than enabled?

File: OpenRA.Mods.Common/UpdateRules/UpdatePath.cs [hunk 2]
```diff
@@ -110,6 +110,8 @@ public class UpdatePath
 				new RemoveResourceExplodeModifier(),
 				new DefineLevelUpImageDefault(),
 				new RemovedAutoCarryallCircleTurnSpeed(),
+				new ReplacedWithChargeAnimation(),
+				new RenameWithResources(),
```
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/UpdatePath.cs]: This seems to be missing `ReplacedWithSiloAnimation`

File: OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs (hunk 1 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs]: Disabling a SpriteBody makes them invisible.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs]: IMO resetting the animation from TraitResumed is a bodge, because pausing/resuming doesn't have any impact on the animation.  What is actually happening is that any animation config done in the constructor is overridden by `WithSpriteBody.TraitEnabled`, and it just so happens that `TraitResumed` runs immediately after that.

A cleaner implementation would be to do something like:

```csharp
public class WithChargeSpriteBody : WithSpriteBody
{
	readonly AttackCharges attackCharges;
	public WithChargeSpriteBody(ActorInitializer init, WithChargeSpriteBodyInfo info)
		: base(init, info, () => 0)
	{
		attackCharges = init.Self.Trait<AttackCharges>();
		ConfigureAnimation(init.Self);
	}

	void ConfigureAnimation(Actor self)
	{
		var attackChargesInfo = (AttackChargesInfo)attackCharges.Info;
		DefaultAnimation.PlayFetchIndex(NormalizeSequence(self, Info.Sequence),
			() => int2.Lerp(0, DefaultAnimation.CurrentSequence.Length, attackCharges.ChargeLevel, attackChargesInfo.ChargeLevel + 1));
	}

	protected override void TraitEnabled(Actor self)
	{
		// Do nothing - we just want to disable the default WithSpriteBody implementation
	}

	public override void CancelCustomAnimation(Actor self)
	{
		ConfigureAnimation(self);
	}
}
```

`WithSpriteBody.CancelCustomAnimation` needs to be made virtual and overridden here to prevent custom animations from breaking the default state once they complete.

File: OpenRA.Mods.Common/UpdateRules/UpdatePath.cs (hunk 2 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/UpdateRules/UpdatePath.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs (hunk 1 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Render/WithChargeSpriteBody.cs]: Done.

**Comment by reaperrr** [comment]
Updated and rebased.

File: OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			DefaultAnimation.PlayFetchIndex(NormalizeSequence(self, Info.Sequence),
+				() => playerResources.ResourceCapacity != 0
+					? ((info.Stages * DefaultAnimation.CurrentSequence.Length - 1) * playerResources.Resources) / (info.Stages * playerResources.ResourceCapacity)
+					: 0);
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			playerResources = newOwner.PlayerActor.Trait<PlayerResources>();
+			ConfigureAnimation(self);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs]: Why is this needed here?

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs]: I was assuming this is necessary to refresh on owner change?

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs]: The animation accesses the resources via the `playerResources` variable, so updating its value is enough.

File: OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+				yield break;
+
+			var anim = new Animation(init.World, image);
+			anim.PlayFetchIndex(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), Sequence), () => 0);
+
+			yield return new SpriteActorPreview(anim, () => WVec.Zero, () => 0, p, rs.Scale);
+		}
+	}
+
+	public class WithSiloSpriteBody : WithSpriteBody, INotifyOwnerChanged
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithSiloSpriteBody.cs]: Sorry to keep on dragging this out, but while giving this one final look over before merging I noticed the naming inconsistency between this and `WithResourceOverlay`.  Can we please rename this trait to `WithResourcesSpriteBody` and `WithResourceOverylay` to `WithResourcesOverlay`? This has an additional bonus of helping towards #6336.

**Comment by reaperrr** [comment]
Updated. Rebasing the individual resource commits properly would have been hard due to the TS silo yaml changes, so I merged the two commits instead (including their update rules).

File: OpenRA.Mods.Common/UpdateRules/Rules/20180923/RefactorResourceLevelAnimating.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+			locations.Clear();
+		}
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			foreach (var wr in actorNode.ChildrenMatching("WithResources"))
+				wr.RenameKey("WithResourceLevelOverlay");
+
+			var siloAnims = actorNode.ChildrenMatching("WithSiloAnimation", includeRemovals: false);
```
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/Rules/20180923/RefactorResourceLevelAnimating.cs]: We need to include removals for at least the trait renaming!

File: OpenRA.Mods.Common/UpdateRules/Rules/20180923/ReplacedWithChargeAnimation.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+					"You may need to disable/remove any previous (including inherited) *SpriteBody traits\n" +
+					"on the following actors:\n" +
+					UpdateUtils.FormatMessageList(locations);
+
+			locations.Clear();
+		}
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var chargeAnims = actorNode.ChildrenMatching("WithChargeAnimation", includeRemovals: false);
```
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/Rules/20180923/ReplacedWithChargeAnimation.cs]: We need to include removals for at least the trait renaming!

**Comment by reaperrr** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7757936afaaa623a26b27e6e402d16247dbd67f3
